core/unix/signal_linux_aarch64.c

```diff
@@ -54,21 +54,123 @@ save_fpstate(dcontext_t *dcontext, sigframe_rt_t *frame)
     LOG(THREAD, LOG_ASYNCH, 1, "\tsp     = " PFX "\n", sc->sp);
     LOG(THREAD, LOG_ASYNCH, 1, "\tpc     = " PFX "\n", sc->pc);
     LOG(THREAD, LOG_ASYNCH, 1, "\tpstate = " PFX "\n", sc->pstate);
+    LOG(THREAD, LOG_ASYNCH, 1, "\n");
+
+    struct _aarch64_ctx *head = (struct _aarch64_ctx *)sc->__reserved;
+    ASSERT(head->magic == FPSIMD_MAGIC);
+    ASSERT(head->size == sizeof(struct fpsimd_context));
+
+    struct fpsimd_context *fpsimd = (struct fpsimd_context *)sc->__reserved;
+    LOG(THREAD, LOG_ASYNCH, 2, "\tfpsr 0x%x\n", fpsimd->fpsr);
+    LOG(THREAD, LOG_ASYNCH, 2, "\tfpcr 0x%x\n", fpsimd->fpcr);
+    reinterpret128_2x64_t vreg;
+    for (i = 0; i < MCXT_NUM_SIMD_SVE_SLOTS; i++) {
+        vreg.as_128 = fpsimd->vregs[i];
+        LOG(THREAD, LOG_ASYNCH, 2, "\tq%-2d  0x%016lx %016lx\n", i, vreg.as_2x64.hi,
+            vreg.as_2x64.lo);
+    }
+    LOG(THREAD, LOG_ASYNCH, 2, "\n");
+
+    if (proc_has_feature(FEATURE_SVE)) {
+        size_t offset = sizeof(struct fpsimd_context);
+        struct _aarch64_ctx *next_head = (struct _aarch64_ctx *)(sc->__reserved + offset);
+        while (next_head->magic != 0) {
+            switch (next_head->magic) {
+            case ESR_MAGIC: break;
+            case EXTRA_MAGIC: break;
+            case SVE_MAGIC: {
+                const struct sve_context *sve = (struct sve_context *)(next_head);
+                LOG(THREAD, LOG_ASYNCH, 2, "\tSVE VL %d bytes\n", sve->vl);
+                ASSERT(sve->vl == proc_get_vector_length_bytes());
+                const unsigned int vq = sve_vq_from_vl(sve->vl);
+                LOG(THREAD, LOG_ASYNCH, 2, "\tVQ %d\n\n", vq);
+                LOG(THREAD, LOG_ASYNCH, 2, "\tSVE_SIG_ZREG_SIZE %d\n",
+                    SVE_SIG_ZREG_SIZE(vq));
+                LOG(THREAD, LOG_ASYNCH, 2, "\tSVE_SIG_PREG_SIZE %d\n",
+                    SVE_SIG_PREG_SIZE(vq));
+                LOG(THREAD, LOG_ASYNCH, 2, "\tSVE_SIG_FFR_SIZE  %d\n",
+                    SVE_SIG_FFR_SIZE(vq));
+                LOG(THREAD, LOG_ASYNCH, 2, "\tsve->head.size %d\n\n", sve->head.size);
+                LOG(THREAD, LOG_ASYNCH, 2, "\tSVE_SIG_ZREGS_OFFSET %d\n",
+                    SVE_SIG_ZREGS_OFFSET);
+                LOG(THREAD, LOG_ASYNCH, 2, "\tSVE_SIG_PREGS_OFFSET %d\n",
+                    SVE_SIG_PREGS_OFFSET(vq));
+                LOG(THREAD, LOG_ASYNCH, 2, "\tSVE_SIG_FFR_OFFSET   %d\n\n",
+                    SVE_SIG_FFR_OFFSET(vq));
+
+                uint64 vdw; /* A vector's doubleword. */
+                int boff;   /* Byte offset for each doubleword in a vector. */
+                for (i = 0; i < MCXT_NUM_SIMD_SVE_SLOTS; i++) {
+                    LOG(THREAD, LOG_ASYNCH, 2, "\tz%-2d  0x", i);
+                    for (boff = ((vq * 2) - 1); boff >= 0; boff--) {
+                        vdw = *((uint64 *)((((byte *)sve) + (SVE_SIG_ZREG_OFFSET(vq, i)) +
+                                            (boff * 8))));
```

Comment:
Can we use the union instead of this loop: the loop has some confusing computations I don't follow.